### PR TITLE
[Model] Make VoxCPM2 generation settings configurable

### PR DESCRIPTION
**Pull request by lolyhop**
## Purpose

Expose VoxCPM2's LocDiT `inference_timesteps` and `cfg_value` through the existing `hf_overrides.voxcpm2_runtime_config`, preserving defaults of 10 and 2.0. Document these alongside the existing `cfg_cutoff_ratio` setting.

Clamp `inference_timesteps` to at least 2: with default CFG-Zero*, one step skips all estimator calls and returns the initial noise unchanged. Reject non-finite `cfg_value` inputs before they reach the solver.

```yaml
engine_extras:
  hf_overrides:
    voxcpm2_runtime_config:
      inference_timesteps: 4
      cfg_value: 1.5
      cfg_cutoff_ratio: 1.0
```

Related to #4032. Uses the runtime-config mechanism introduced in #3882.

## Test Plan

CPU regression coverage checks parsing, defaults, timestep lower bounds, and non-finite CFG rejection. Solver tests exercise timestep inputs -3, 0, 1, 2, 4, and 10, batch sizes 1 and 3, CFG 1.5 and 2.0, and cutoff 0 and 1. They assert estimator call counts, finite non-noise output, unchanged input noise, and an analytical output for full CFG.

Run in an environment with matching vLLM/vLLM-Omni dependencies (no model weights or GPU required for these tests):

```bash
pytest -q tests/model_executor/models/voxcpm2/test_runtime_config.py -m 'core_model and cpu'
```

**vLLM Version:** Not installed in the local macOS validation environment.

**vLLM-Omni Commit:** Fix based on PR head `d5c01c0b654545c35d6fc670bd74d401517d966b`.

## Test Result

- Isolated CPU validation with Python 3.12.12 / PyTorch 2.14.0: 62 passed. The harness loads production runtime-config and solver definitions directly from source, bypassing vLLM imports; the estimator is a deterministic test module. This is not a full vLLM integration run.
- The same tests using the pre-fix runtime config: 30 failed, 32 passed, reproducing the boundary bug and missing finite-value validation.
- Ruff check, Ruff format check, and `git diff --check`: passed.
- Repository pytest collection is blocked locally by missing `vllm`. CUDA Graph replay and synthesis-quality evaluation have not been run.

### CODE REVIEW

**Comment by Sy0307** [comment]
If lower inference timesteps will affect quality? Have you tested it?

**Comment by lolyhop** [comment]
> If lower inference timesteps will affect quality? Have you tested it?

The change preserves the original default of `inference_timesteps=10`, so it does not affect quality unless the value is explicitly overridden.

In general, reducing the number of steps may degrade synthesis quality, but it can improve inference latency and throughput. This trade-off can be useful for latency-sensitive deployments, as well as for distilled LocDiT variants that are trained to produce good results with a different or smaller number of solver steps. 

The purpose of this PR is to make that experimentation and deployment possible without changing the default behavior.

**Comment by Sy0307** [comment]
I understand your point, and I know this still keeps the default setting. However, I'm concerned that making it optional might sometimes confuse users, and certain agents could adjust it incorrectly. I think it would be best to add some comments here as a warning to better alert users. Just for your reference.

**Comment by lolyhop** [comment]
> I understand your point, and I know this still keeps the default setting. However, I'm concerned that making it optional might sometimes confuse users, and certain agents could adjust it incorrectly. I think it would be best to add some comments here as a warning to better alert users. Just for your reference.

Okay, good point. I’ve added a concise warning.

File: vllm_omni/model_executor/models/voxcpm2/runtime_config.py [hunk 1]
```diff
@@ -111,7 +113,12 @@ def _coerce_value(key: str, value: Any, default: Any) -> Any:
             return bool(value)
         if isinstance(default, int) and not isinstance(default, bool):
             value = int(value)
-            if key in {"audio_emit_every", "vae_decode_every", "batched_fsq_fusion_max_batch"}:
+            if key in {
+                "inference_timesteps",
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/voxcpm2/runtime_config.py]: [P2] Prevent a timestep count that skips all estimator calls

Adding `inference_timesteps` to this clamp permits `1` and normalizes non-positive inputs to `1`. With the default `use_cfg_zero_star=True`, `_optimized_solve_euler_with_noise` sets `zero_init_steps=1` and `start_step=2`; for `n_timesteps=1`, `len(t_span)=2`, so the loop is empty and the solver returns its input noise unchanged.

Reproduced with the production solver on CPU using a counting estimator: inputs `-3`, `0`, and `1` all normalize to one step, make zero estimator calls, and return the original noise unchanged. Inputs `2`, `4`, and `10` make `1`, `3`, and `9` estimator calls, respectively.

Please require at least two steps with the current solver, or explicitly implement one-step support, and add a solver-level regression test. The new parsing tests currently accept this broken boundary.

**Review by hsliuustc0106 — CHANGES_REQUESTED** [review]
The configuration plumbing preserves the current defaults and reaches the ordinary, batched, and CUDA graph generation paths. I found one correctness issue with the newly permitted minimum timestep count, detailed inline.

Validation: the five added configuration cases passed in isolation with the vLLM logger stubbed; YAML parsing and diff whitespace checks passed. I also reproduced the minimum-step issue by executing the production solver and buffer definitions with actual CPU tensors and a counting estimator. The normal repository pytest run was blocked by a broken NumPy/PyTorch import in the review environment. Full-model synthesis and GPU validation were not run.

File: vllm_omni/model_executor/models/voxcpm2/runtime_config.py (hunk 1 above)
Comment (lolyhop) [reply vllm_omni/model_executor/models/voxcpm2/runtime_config.py]: Good catch, thanks! Fixed the minimum to 2 and added solver-level regression tests covering this boundary.

File: vllm_omni/model_executor/models/voxcpm2/runtime_config.py [hunk 2]
```diff
@@ -111,10 +114,14 @@ def _coerce_value(key: str, value: Any, default: Any) -> Any:
...
         if isinstance(default, int) and not isinstance(default, bool):
             value = int(value)
+            if key == "inference_timesteps":
+                return max(2, value)
             if key in {"audio_emit_every", "vae_decode_every", "batched_fsq_fusion_max_batch"}:
                 return max(1, value)
             return value
         if isinstance(default, float):
+            if key == "cfg_value" and not math.isfinite(float(value)):
+                raise ValueError("VoxCPM2 cfg_value must be finite.")
```
Comment (linyueqian) [inline vllm_omni/model_executor/models/voxcpm2/runtime_config.py:124]: [suggestion] `cfg_value` is only checked for finiteness, while its neighbour `cfg_cutoff_ratio` is clamped to `[0.0, 1.0]` two lines below.

That leaves a negative `cfg_value` accepted. Classifier-free guidance with a negative scale does not fail, it inverts: the sampler moves away from the conditioned prediction, so the result is intelligible-sounding audio that ignores the prompt, which is harder to diagnose from a deployment than a startup error would be.

A `max(0.0, ...)` floor, or an explicit raise for negatives alongside the existing non-finite raise, would make the two settings behave consistently. Not blocking, since the default is unchanged and anyone setting this is opting in deliberately.

**Review by linyueqian — APPROVED** [review]
Approving. This is the right shape for making a hardcoded knob configurable: the two values move into `_VoxCPM2RuntimeConfig` with defaults identical to what was inlined at `voxcpm2_talker.py:880-881`, so a deployment that sets nothing behaves exactly as before, and the yaml carries them with a comment saying to leave them alone unless validated. The 116 lines of tests cover the coercion paths rather than just the happy path.

The validation is where this kind of change usually goes wrong, and it mostly does not here. `inference_timesteps` floors at 2 in the same style as the existing `max(1, ...)` keys, and rejecting a non-finite `cfg_value` outright is right, since NaN would propagate silently through guidance rather than failing.

One asymmetry left, noted inline: `cfg_cutoff_ratio` is clamped to a range while `cfg_value` only has to be finite, so a negative value is accepted and inverts guidance rather than being rejected or clamped.

CI: `buildkite/vllm-omni` green at this head, NPU green. Intel and AMD carry the standing inherited reds, so `ci_approval_override_head` is recorded at this head for those two.

Validation: static review of the five-file author diff against merge base, the coercion branch in `_coerce`, and the call site the defaults replaced. Fork head, so the branch was not executed.

**Comment by linyueqian** [comment]
@hsliuustc0106 your change request from 2026-09-06 looks resolved at the current head, so this is only blocked on a re-review.

You asked for at least two steps with the current solver, having reproduced that `-3`, `0` and `1` all normalize to one step, make zero estimator calls, and return the input noise unchanged. `_coerce` now returns `max(2, value)` for `inference_timesteps` at `runtime_config.py:117-118`, so every one of those inputs clamps to 2 and the `t_span` degenerate case you described is unreachable through configuration.

I should have said this in my own review rather than approving alongside an open request and leaving you no signal, which is why this sat. Flagging it now so the merge is not waiting on a finding that has already been addressed.

The one point from your comment still worth a decision is the regression test: the parsing tests cover the new boundary, but there is still no solver-level test asserting that a configured minimum actually produces estimator calls. That is a smaller ask than the original finding and does not have to block.

**Comment by lolyhop** [comment]
@hsliuustc0106, could you take another look, please?

**Review by hsliuustc0106 — APPROVED** [review]
